feat(workspaces): Integrating BE with FE

foochifa opened the pull request:
## Problem
<!-- What problem are you trying to solve? What issue does this close? -->
Integrating the frontend with all the workspaces BE works that @hanstirtaputra have done

Context can be found in #4222 

## Solution
<!-- How did you solve the problem? -->

Added workspace service calls and useQueries.

Integrated workspace components and properly use `WorkspaceContext` from workspace searching and filtering done by @justynoh #5458 

Also fixed BE call for form delete to also delete from workspaces

**Breaking Changes** 
<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? -->
- [ ] No - this PR is backwards compatible  to workspaces branch only

## Before & After Screenshots

**BEFORE**:
<!-- [insert screenshot here] -->

**AFTER**:
<!-- [insert screenshot here] -->

## Tests
<!-- What tests should be run to confirm functionality? -->

foochifa committed:
feat: add api services for workspace

foochifa committed:
feat: update Workspace page with default workspace

foochifa committed:
feat: update workspace context

foochifa committed:
feat: add workspace mutations

foochifa committed:
chore: trim workspace content props

foochifa committed:
feat: add functionality for rename and delete workspace

foochifa committed:
feat: add default workspace to menu

foochifa committed:
feat: implement create workspace

foochifa committed:
feat: update workspace header

foochifa committed:
feat: implement row action and move form to workspace

foochifa committed:
feat: remove workspace mocks from browser

foochifa committed:
feat: remove form from workspaces after archive

foochifa committed:
feat: reset rename modal useform value

foochifa committed:
fix: do not mutate if title is same

foochifa committed:
fix: move to selected workspace even form exists

foochifa committed:
fix: do not switch to workspace after moving

foochifa committed:
fix: mobile mode for workspace

foochifa committed:
chore: clean unused attributes

foochifa committed:
chore: improve spacing for desktop workspace header

foochifa committed:
chore: improve spacing for mobile view too

foochifa committed:
Merge branch 'develop' into form-v2/workspace-integrate-fe

foochifa committed:
fix: align workspace with develop

foochifa committed:
fix: useQuery keys

foochifa committed:
Merge branch 'form-v2/form-workspaces' into form-v2/workspace-integrate-fe

foochifa committed:
feat: delete workspace message

foochifa commented:
Will fix test cases but welcome for review first

Also Note: Pushing this to a feature branch and not develop yet

foochifa commented:
@kenjin-work https://www.figma.com/file/CMq0CJtawn7NGM7Zdpi8Ti/Form-Design-Master?type=design&node-id=185-118&mode=design&t=ldzjPrkjHCZRbSMk-0

kenjin-work commented:
UI review first:

There's a small gap on the right that I don't see in the Figma:
![image](https://github.com/opengovsg/FormSG/assets/119096102/27ad641f-bfd3-4025-bbdb-e0d8e2c86370)

No tick on the current workspace that the form is in (in this case this form's belongs to `test` but it's not showing)
![image](https://github.com/opengovsg/FormSG/assets/119096102/531194fa-faee-410b-a808-0287cfe29954)

Backend (can be addressed in another PR):
Workspaces with duplicate names can be created. When admin selects workspace to move to this can lead to confusion.

kenjin-work commented on frontend/src/features/workspace/components/WorkspaceFormRow/RowActions/RowActionsDropdown.tsx:
```diff
@@ -1,26 +1,73 @@
+  const { handleMoveForm } = useRowAction(formMeta)
+  const { workspaces } = useWorkspaceContext()
+
+  const handleWorkspaceClick = (destWorkspace: Workspace) =>
+    handleMoveForm(destWorkspace._id.toString(), destWorkspace.title)
```
Should use a `useCallback` here (might be more efficient).

foochifa replied:
Thanks @kenjin-work! [ad5b50e](https://github.com/opengovsg/FormSG/pull/6560/commits/ad5b50e7937face50761c8b0d1f84f68d3ac0c4c)

kenjin-work reviewed (commented):
One smol comment for code. Mostly LGTM.

foochifa committed:
fix: BE tests

foochifa committed:
fix: delete workspace copy

foochifa committed:
chore: abstract out use context form workspace modals

foochifa committed:
fix: mock workspace for storybook

foochifa committed:
fix: mobile workspace page grid and menu padding

foochifa committed:
feat: display checkbox on form's current workspace

foochifa committed:
feat: use callback and memo for workspace row actions

foochifa committed:
feat: mobile drawer actions

foochifa commented:
Heya @KenLSM @LinHuiqing 

Kenjin reviewed this before he left, and have addressed his comments. 

However, i have made certain changes since and would like another pair of eyes!

1. To show current workspace for form, have made a util method within the context to find the workspace the form belongs in. Will then populate `RowActionDrawer` or `RowActionDropdown`
2. Implemented `RowActionDrawer` for mobile view and fixed some grid issues to ensure mobile is okay

Let me know if you need more context (Will also update the manual test section)! After this, there's only 1 more issue left which is to delete form from workspace when removed as collaborator!.

foochifa committed:
fix: move emptyworkspace to WorkspaceContent

KenLSM reviewed (approved):
LGTM!
